UI Filters (Implementation)

alex40724 opened the pull request:
This is a follow up on #1193 which needed rebasing. The old PR included the toggle button and other cherry-picks which made it hard to read.

Open (at least)

- [x] Writing tests
- [x] Cleanup less

There is some similar code in forms and filters now. Most obviously `PostDataFromServerRequest` which could be moved one directory level up. The `FilterContextRenderer` is another example. We could introduce some traits here, but personally I am not sure how much this will differ in the future.

alex40724 committed:
cp: First draft for Filters, mostly copied from Forms for the beginning

alex40724 committed:
cp: Second draft, more cleaned up

alex40724 committed:
cp: Ready for Pull Request

alex40724 committed:
cp: Wording

alex40724 committed:
cp: Changes for Filter Field

alex40724 committed:
cp: fixed some rules

alex40724 committed:
cp: very, very basic filter ideas

alex40724 committed:
cp: Changes from comments and JF

alex40724 committed:
cp: Trifle

alex40724 committed:
cp: Trifle again

alex40724 committed:
cp: Still basic Filter Service ideas

alex40724 committed:
cp: Still basic Filter Service ideas 2

alex40724 committed:
cp: sketched processing

alex40724 committed:
cp: First try to get data from request

alex40724 committed:
cp: merged filter changes from 5_4_cat_filters2 up to b411ae1 (without toggle button changes)

alex40724 committed:
cp: merged filter changes from 5_4_cat_filters2 up to 8820e63

alex40724 committed:
cp: merged filter changes from 5_4_cat_filters2 up to 9198d37

alex40724 committed:
cp: Basic functionality of showing Labels and Values in the Filter Bar +...

alex40724 committed:
cp: First ideas for Add-Button and Remove-Glyph

alex40724 committed:
cp: merged changes from 5_4_cat_filters2 of c872761

alex40724 committed:
updated style json, composer autoloader and delos.css

alex40724 committed:
use correct glyphs

alex40724 committed:
accordion behavior for popovers in mobile view

alex40724 committed:
force expanded if activated and non-mobile

alex40724 committed:
fixed form submission

alex40724 committed:
add select

alex40724 committed:
add select

alex40724 committed:
fixed value being displayed in proxy

alex40724 committed:
fixed session handling for values

tfamula committed:
Behavior for removing and adding Inputs to the Filter

tfamula committed:
Merge branch '5_4_ks_filter2' of https://github.com/leifos-gmbh/ILIAS into 5_4_ks_filter2

alex40724 committed:
fixed getting data from service

alex40724 committed:
Merge branch '5_4_ks_filter2' of https://github.com/leifos-gmbh/ILIAS into 5_4_ks_filter2

tfamula committed:
Removed double inheritance

tfamula committed:
Merge branch '5_4_ks_filter2' of https://github.com/leifos-gmbh/ILIAS into 5_4_ks_filter2

alex40724 committed:
restored tabindex

tfamula committed:
Accessibility for Input Fields and improvement for adding Input Fields

alex40724 committed:
show/hide filter behaviour

alex40724 committed:
Merge branch '5_4_ks_filter2' of https://github.com/leifos-gmbh/ILIAS into 5_4_ks_filter2

alex40724 committed:
polished ui service

alex40724 committed:
updated readme

alex40724 committed:
polished renderer

alex40724 committed:
some less cleanup, more needed, and variables

alex40724 committed:
some less, js polishing, needs more

alex40724 committed:
fixed some type hints

alex40724 committed:
moved action generating to request adapter

alex40724 committed:
added missing mem vars

alex40724 committed:
added missing mem vars

alex40724 committed:
added missing purpose/rules

Amstutz commented:
@alex40724 Is this ready for a final review? Keep in mind, that we will need some time to review this, due to the raw amount of changes. Ping me if ready. However, this is a key component and I know that @klees will want to review the final implementation for this as well.

alex40724 commented:
@Amstutz Yes, this is ready for final review, except that we are working on less and tests. And of course we are still find things that we improve (missing type hints and stuff). But all the main code is ready for review.

tfamula committed:
Less variables + cleanup

alex40724 committed:
get rid of DIC dependency

tfamula committed:
Started Filter tests

Amstutz commented on Services/UI/classes/class.ilUIFilterRequestAdapter.php:
```diff
@@ -0,0 +1,105 @@
+	 * @param \ILIAS\UI\Component\Input\Container\Filter\Standard $filter
+	 * @return \ILIAS\UI\Component\Input\Container\Filter\Standard
+	 */
+	public function getFilterWithRequest(\ILIAS\UI\Component\Input\Container\Filter\Standard $filter): \ILIAS\UI\Component\Input\Container\Filter\Standard
```
Why not using the "use" statement in the very top, would make this a bit more readable

Amstutz commented on src/UI/Implementation/Component/Input/Container/Filter/Filter.php:
```diff
@@ -0,0 +1,317 @@
+	/**
+	 * Check the request for sanity.
+	 *
+	 * TODO: implement me!
```
Do it.

Amstutz commented on src/UI/Implementation/Component/Input/Container/Filter/Filter.php:
```diff
@@ -0,0 +1,317 @@
+	}
+
+
+	// Implementation of NameSource
```
Strange comment format.

Amstutz commented on Services/UI/README.md:
```diff
@@ -0,0 +1,46 @@
+		return $renderer->render($filter)."Filter Data: ".print_r($filter_data, true);	
+	}
+
+}
```
Thx for the example in the readme.

Amstutz commented on Services/UI/classes/class.ilUIFilterService.php:
```diff
@@ -0,0 +1,216 @@
+	 *
+	 * @param string $filter_id
+	 * @param string $base_action
+	 * @param array $inputs
```
declare the type of the array here, it's an array of ILIAS\UI\Implementation\Component\Input\Field, correct?

klees replied:
```suggestion
	 * @param ILIAS\UI\Component\Input\Field\FilterInput[] $inputs
```

klees commented on Services/Init/classes/class.ilInitialisation.php:
```diff
@@ -1579,7 +1579,13 @@ protected static function initUIFramework(\ILIAS\DI\Container $c) {
+						  	, $c["ui.template_factory"]
+						  	, $c["lng"]
+						  	, $c["ui.javascript_binding"]
+						  )
```
```suggestion
							)
```

klees commented on Services/UI/classes/class.ilUIFilterService.php:
```diff
@@ -0,0 +1,216 @@
+							 $is_activated = false, $is_expanded = false): \ILIAS\UI\Component\Input\Container\Filter\Standard
+	{
+		global $DIC;
+		$ui = $DIC->ui()->factory();
```
Why do you inject other dependencies but not this one?

alex40724 replied:
done

klees commented on Services/UI/classes/class.ilUIFilterService.php:
```diff
@@ -0,0 +1,216 @@
+	 * @param string $filter_id
+	 * @param string $base_action
+	 * @param array $inputs
+	 * @param array $is_input_initially_rendered
```
```suggestion
	 * @param bool[] $is_input_initially_rendered
```

klees commented on Services/UI/README.md:
```diff
@@ -0,0 +1,46 @@
+		$filter_data = $DIC->uiService()->filter()->getData($filter);
+	 
+		//Step 4: Render the filter
+		return $renderer->render($filter)."Filter Data: ".print_r($filter_data, true);	
```
This won't render the values that the user entered imo. They are added from POST in `getData`, but the updated `$filter` is never retrieved. Thus `$filter` here only contains the values that where initially set to the fields.

alex40724 replied:
![bildschirmfoto 2018-10-29 um 10 39 30](https://user-images.githubusercontent.com/1221841/47641701-09f58580-db67-11e8-9940-ddb505b69127.png)

What do you mean "won't" render? Did you try the example in the KS? You enter some data in the "Title" field, hit "Apply" (Do not use "Enter" key, this needs fixing) and the data is retrieved and displayed.

klees replied:
I just now noticed, that you effectively add the POST-data to the input twice, once in `handleApply`, once in `getData`. This was irritating but should work.

klees commented on Services/UI/classes/class.ilUIFilterService.php:
```diff
@@ -0,0 +1,216 @@
+			foreach ($filter->getInputs() as $input_id => $i)
+			{
+				$this->session->writeValue($filter_id, $input_id, $i->getValue());
+			}
```
As far as I could follow the code the `getInputs` basically gets the list of inputs from the `InputGroup`that forms the `Filter Container`. This will just be a list, so there are no "input_ids" here, just positions in the list. So this might or might not work, no guarantees about stable ids here.

alex40724 replied:
Yes, I see. Would it be ok, to make the non numeric keys in the submitted array mandatory? Like in the example?
```
[
		"title" => $title_input,
		"select" => $select,
		"with_def" => $with_def,
		"init_hide" => $init_hide,
	]
```
We could test for these keys. This would fix this issue for now.

klees replied:
No, there just are no guarantees here. This will work for now, but since the inputs are not meant to be read in this way this could well change. This is what I meant with "against the grain" in my other comment.

klees commented on src/UI/Implementation/Component/Input/Field/FilterContextRenderer.php:
```diff
@@ -0,0 +1,293 @@
+		$tpl->setVariable("POPOVER", $default_renderer->render($popover));
+
+		$prox = new ProxyFilterField();
+		if ($input->getByline() != "disabled")
```
What's this magic string here? To me this looks as if you use the byline to pass other information here.

alex40724 replied:
Yes, exactly. Unfortunately we do not have disabled inputs yet. I think we need this for forms as well in the future. I would mark this as a workaround since there is no other way to get this info through (as far as I know). The bylines are currently not needed. If there is currently a better way to achieve this (with the current interfaces) I would be happy do change this.

klees commented on src/UI/Implementation/Component/Input/Field/FilterContextRenderer.php:
```diff
@@ -0,0 +1,293 @@
+							il.UI.filter.onRemoveClick(event, '$id');
+							return false; // stop event propagation
+					});";
+				//var_dump($code); exit;
```
No uncommented code please.

alex40724 replied:
has been removed

klees commented on src/UI/Implementation/Component/Input/Field/FilterContextRenderer.php:
```diff
@@ -0,0 +1,293 @@
+			$inputs .= $default_renderer->render($input);
+			$input_labels[] = $input->getLabel();
+		}
+		if ($group->getLabel() != "disabled")
```
What's this magic string here? To me this looks as if you use the label to pass other information here.

alex40724 replied:
See https://github.com/ILIAS-eLearning/ILIAS/pull/1253#discussion_r228858283

alex40724 commented:
@Amstutz We are done with less.
@tfamula Introduced lots of less variables.

alex40724 committed:
fixed duplicate class name

klees reviewed (commented):
Hi @tfamula and @alex40724!

Please note, that my review is not complete yet. I just spend two hours reviewing, but there's a lot to see here. I will proceed at noon.

I left some inline comments and I already have three general remarks. I might be missunderstanding something, though:

* This seems to go against the grain of how we designed the inputs. General idea of the inputs is to use a flow like: use current data to build a container using withValue, attach transformations to the inputs that retrieves the desired datastructure; attach post data to that container; render result, retrieve updated data structure. The usage here on the other hand needs intimite knowledge of the inner workings of the filter, most visible due the fact that `getInput` needs to be made public on the `Filter`.
* I think we do not speak proper HTTP if we use POST for passing the values of the filter. This already shows e.g. in thecurrent  search of ILIAS, where the browser (at least mine) explicitely prompts for resubmitting the form when just reloading the page because he suspects that the data I submitted gets stored or is processed in some other way on the server. For filters, on the other hand, the query-parameters seem to be the obvious HTTP-feature to use. I would be very happy if we did this right for a new filter implementation.
* I'm not very confident that this implementation works reliably yet. The API you show in the REAME will not display the values that the user submitted (or I am missing something crucial). The renderer implementation seems to rely on magic strings in the label and byline field. The data storage in the session seems to rely on guarantees that the implementation of the inputs just doesn't provide. The connection between single inputs in the filter and the information whether they should be shown is somehow loose (two separate lists). This seems to related to the first remark. I might be missing some subleties, though.

I know that my remarks imply fundamental changes and I will be happy to help to find proper solutions. Sadly there is very little time for such a huge and important feature until the release needs to be ready.

alex40724 committed:
cherry picked 06871bf4bb5ad38473ef2be5f61c90db3f09ba1a from loading animation branch

alex40724 committed:
fixed more tests

klees committed:
Update Services/Init/classes/class.ilInitialisation.php

Co-Authored-By: alex40724 <alex.killing@gmx.de>

klees committed:
Update Services/UI/classes/class.ilUIFilterService.php

Co-Authored-By: alex40724 <alex.killing@gmx.de>

klees committed:
Update Services/UI/classes/class.ilUIFilterService.php

Co-Authored-By: alex40724 <alex.killing@gmx.de>

tfamula committed:
Added some more tests

alex40724 committed:
fixed tests; removed vardump

alex40724 commented:
@klees 

> The connection between single inputs in the filter and the information whether they should be shown is somehow loose (two separate lists). This seems to related to the first remark.

This point has been discussed extensively during JF and been accepted in the interface.

> For filters, on the other hand, the query-parameters seem to be the obvious HTTP-feature to use.

This sounds feasible. I will not be able to move this from POST to GET today. I would have been nice to have this feedback two weeks earlier. The code was based on POST handling since the beginning.

> This seems to go against the grain of how we designed the inputs. General idea of the inputs is to use a flow like: use current data to build a container using withValue, attach transformations to the inputs that retrieves the desired datastructure; attach post data to that container; render result, retrieve updated data structure. The usage here on the other hand needs intimite knowledge of the inner workings of the filter, most visible due the fact that getInput needs to be made public on the Filter.

This I do not understand. The procedure is very similar to the forms. The consumer calls `getData` very similar to forms, the data structure is the same and `getInputs` is public in the Forms as well (the renderer needs this).

alex40724 committed:
pass ui as dependency

alex40724 committed:
first tests using get

Amstutz commented on src/UI/Implementation/Component/Input/Field/FilterContextRenderer.php:
```diff
@@ -0,0 +1,292 @@
+		$tpl = $this->getTemplate("tpl.context_filter.html", true, true);
+
+		if ($input->getByline() == "disabled") {
+			$remove_glyph = $f->glyph()->remove()->withUnavailableAction(true);
```
I do no understand that. Why those this happen here?

Amstutz commented on src/UI/templates/js/Input/Container/filter.js:
```diff
@@ -0,0 +1,178 @@
+
+	$("*").on("il.ui.popover.show", function(e){
+		//
+	});
```
What is this doing, why?

Amstutz commented on src/UI/templates/js/Input/Container/filter.js:
```diff
@@ -0,0 +1,178 @@
+		});
+	};
+
+	$(init);
```
Why is this placed here, what is different from this code here, than from the code bellow in: (document).ready(function() ) on the very botom in this file. I am sure there is something I am missing here.

Amstutz commented on src/UI/templates/js/Input/Container/filter.js:
```diff
@@ -0,0 +1,178 @@
+
+			//Show labels and values in Filter Bar
+            var input_name = $el.attr("name");
+            var input_num = input_name.substring(13);
```
I do not undersand this line here. What is the purpose of this?

Amstutz commented on src/UI/templates/js/Input/Container/filter.js:
```diff
@@ -0,0 +1,178 @@
+            var addableInputs = $el.parents(".il-standard-form").find(".il-popover-container:hidden").length;
+            if (addableInputs != 0) {
+                $("#" + id).parents(".il-standard-form").find(".btn-bulky").parents(".il-popover-container").show();
+            }
```
Thanks for the comments in this part. this helps.

Amstutz commented on src/UI/templates/js/Input/Container/filter.js:
```diff
@@ -0,0 +1,178 @@
+            }
+
+            //Hide the Popover of the Add-Button when adding Input Field
+			$el.parents(".il-standard-form").find(".btn-bulky").parents(".il-popover-container").find(".il-popover").hide();
```
Thanks for the comments.

Amstutz reviewed (changes requested):
Hi @tfamula and @alex40724,

Thanks for your tremandeous work on this PR. We see that you spend many hours here. We also see that you took care of ensuring good code quality (by adding comments and using proper docs). Please note that we did not take the decision lightly to not merge this PR at this point.

Please note as well, that our [contributor guidelines](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/docs/documentation/contributing.md) state certain [demands to contributions](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-3/docs/documentation/contributing.md#rules-for-contributors) in this case most notably that huge PRs should be discussed with responsible maintainers in advance and "that your code could be understood and is documented - this will help reviewers as well".

This implementation PR came in on friday (effectively one workday ago) and this left us very little time to review and understand your changes. We understand that you are chasing deadlines but neither the contribution rules nor the "Filter"-feature request or the shared understanding that the UI Framework should adhere to certain standards is news to anyone. We both still took substantial amounts of time today to review this PR. Unfortuntly we see certain issues in this PR that require further work (from you and us) before we consider this to be mergeable to the UI Framework. Note that such issues can be avoided by holding a workshop with the responsible maintainer early in the development process.

We are sorry that this most probably will mean that this contribution misses the deadline for 5.4. We do not think, though, that the short term goal of the release should impede our longterm goal of stable and reliable code in the UI framework in this way, especially because these requirements and the filter are not new. 

However, if you manage to clear all those points before the deadline or if we find another solution that maintains our goal of stable and reliable code in the UI-framework, we are of course willing to merge this.

We found the following core issues while reviewing the filter in their current state: 
* Even after a longer review, we would have trouble to add a new Input Field as Input Field for filters because we currently do not understand what is necessary and why. Example: What would we need to do, the make Numeric Inputs for filters? Does it suffice to extend the FilterInput? Please add an according section to the Readme Input Handling in the UI-Framework. Further, we believe the way to get new inputs into the filter is very far. 
* The handling of the disabled state of fields is problematic. It is done by using a magic string in the byline or label field. We see the reason why it is done this way, but we would suggest a disabled attribute on the interface of the filter instead of using this kind of magic. If this is not possible due to some reason, at least communicate clearly in comments and Todos, why it is done this way. We then would expect a PR from you asap that fixes the underlying issue.
* The current js to us is cryptic at the moment. We consider it crucial, that we understand what it does and why, because only this will allow us to take our coordinators responsibility in this context. Please leave some explanations and comments or, better, factor out some methods to allow us to see some structure in the client-side code (see comments).

Especially the fact, that we do not understand crucial parts of the PR (JS, adding of inputs to the filter) currently leaves us in state of mind where we cannot, with clear conscience, say: yes, we think this implementation is good for the UI framework. In other similar cases (e.g. the first 2 iterations of the modal implementation with the new signal handling) we rejected the first proposals and asked for further explanations and revisions. For the forms we ourselves took at least 3 major iterations before having a solution we believed to be stable, reliable and extendable enough to be shipped. The solution does not have to be perfect, but we need to understand the implications at least. For this we will require solid documentation of you approach to form an understanding. Note that such complex components need their time and iterations before they are really good. However they need to be, since once introduced it will be very hard to change them retroactively.

Further we found the following issue we consider important
* The tests are highly incomplete (e.g. where are the rendering tests)?
* Currently, the raw values are saved in the session of your service. This is not an issue of your UI-Components related  code, however it is against the principle of how constraints are attached to inputs. We believe a service that highly coupled to the UI service should respect this principles. What we believe what should be done, is to implement some mapping taking care of coupling the content of the filter (stored in sessions, secured by constraints etc.) to the values that are shown to the users. Currently the raw values of the users are stored directly in the session (and therefore DB). They thus are not subject to security mechanisms implemented in the inputs of the UI framework because they currently bypass implemented means.
* The current implementation of the Filter via POST doesn't seem to be proper HTTP. Although the current filters in ILIAS do the same thing, using GET with query-params for the filter seems to be correct here. This could be a task for the next release, though.

We also found aspects, where your implementation really shines:
* We like the way you implemented the scoping/encapsulation of the JS
* The you propose a good choice of variables for the less code.

Finally there are minor issues pointed out directly in the comments.

If all those things are not feasible anymore for 5.4, note that we would like to see them for 6.0 and will both cooperate with time, brain and code as required. Also note that especially a possible revised table implementation for the UI Framework there could be funding available to cary the filter components as well over this final stretch. So even if time runs out for this for 5.4, this is not lost effort. 

Kindly, @klees and @amstutz

tfamula committed:
Render tests

tfamula committed:
Merge branch '5_4_ks_filter2' of https://github.com/leifos-gmbh/ILIAS into 5_4_ks_filter2

tfamula committed:
Fix for render tests

tfamula committed:
POST tests

tfamula committed:
Second tests using GET

tfamula committed:
GET not working actually

tfamula committed:
GET with javascript

tfamula committed:
GET with javascript 2

tfamula committed:
Bugfix for filter request

tfamula committed:
JS cleanup

tfamula committed:
Removed double inheritance for Radio Input

tfamula committed:
Interface, implementation and examples for disabled inputs

tfamula committed:
Prevent request for disabled inputs

tfamula committed:
Disabled state for (dependant) group and section

tfamula committed:
Temporary template change to prevent conflict with cherry-pick

tfamula committed:
Restoring template changes

tfamula committed:
Filter deactivation with real disabled inputs

tfamula committed:
Cherry pick from disabled input PR

tfamula committed:
Changes for disabled Tag Input

alex40724 commented:
@Amstutz @klees 

We took care of most the crucial points you mentioned

- Moved from POST to GET handling
- Refactored the disabled state (see separate PR)
- Worked on the JS

We will continue to work on

- the documentation and 
- the tests

The main open point for us is

>  Currently, the raw values are saved in the session of your service. ...

We need to feed the "raw" values from the sessions back into the inputs via withValue. There is no way to do this with the content we get after the operations have been processed. This is what I asked Timon some time ago and he answered that there is no bijectivity in these operations (e.g. if it operates "a + b = c" I will not get the a, b values out of a "c". So I cannot put "c" into the session, since it is no usable for representing the filter which needs "a" and "b".

It is up to the DB/Session layer to escape any(!) values wrt security. The UI service feeds these values back into the same process (withValue) as the forms do when using withPostRequest (which is direct insecure POST data). So there is no big difference wrt to security imo.

tfamula committed:
Removed comment

tfamula committed:
Added documentation and fixed tests

tfamula committed:
Solved violation with immutability

Amstutz commented:
@alex40724 , @tfamula @klees and @Amstutz discussed in a skype session the following points:
- How do we persist filter (Raw Values in Session)?
  -  Decision: We accept the current implementation of using a Filter Service to take care of this task
    - Trade Off: The current implmentation will make it harder to implement reusable filter parts
    - Advantage: The Filter will be easier to use.
- Do we need On-Update on all Inputs?
  - Decision: We decided to try to introduce this on all fields and so get rid of the filter interface.
- Rendering-Context of Filter Inputs
  - We decided that the current implementation is fine, but that we have to oberve this carefully since we do not want to have a great difference between the rendering in filters and their regular version.
- JS readability and dependencies?
  - There is a dependency from the name source on the PHP-Side to JS, we consider that to be a problem.
  - The readability of the JS code still need to be enhanced.

We agreed that the following points have to be tackled before merging:
- [x] Dissolve dependency from NameSource to JS
- [x] Propose OnUpdate on all Inputs
- [x] Further factor out JS-Code to more individual functions to increase the readabiliy of the code.

tfamula committed:
Restructuring JS Code

tfamula committed:
Fix to prevent losing filter status when clicking on collapse without apply

tfamula committed:
Removed too much code

Amstutz commented:
Just making sure that we are not stuck in a deadlock here, we will get some little notch, as soon as this is ready for a final review, correct?

tfamula committed:
OnUpdate for all inputs

tfamula commented:
Hi @Amstutz and @klees!
We tackled the open points which came up in the skype session. This PR is ready for a final review now. We look forward to your feedback next year!

Amstutz commented:
Hi @tfamula thanks for the improvements, this is still on my radar, I will perform the review untill the end of next week the latest.

klees commented:
Same here!

klees commented on src/UI/Component/Input/Container/Filter/Factory.php:
```diff
@@ -0,0 +1,47 @@
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *      The standard filter is the default filter to be used in ILIAS. If there is no good reason using another filter instance in ILIAS, this is the one that should be used.
```
```suggestion
	 *      The standard filter is the default filter to be used in ILIAS. If there is no good reason
	 *      using another filter instance in ILIAS, this is the one that should be used.
```

klees commented on src/UI/Implementation/Component/Input/Container/Filter/Filter.php:
```diff
@@ -0,0 +1,343 @@
+	protected $input_group;
+
+	/**
+	 * @var array
```
```suggestion
	 * @var bool[]
```

Amstutz reviewed (changes requested):
Hi @tfamula
Thanks a lot for your great stamina on this topic. This is complex however, we are getting there.
My review is mainly based on some basic checks I performed on your example, making sure, the behaviour of the implementation, matches the given the description. @rklees might leave some more comments about the more technical aspects of this. 

Please answer the following questions:
- [ ] What is the relation between collapse and activate? From your description I gather: 

> An additional row with set filter settings will be shown at the bottom of the Filter Bar, when the Filter Bar is collapsed and the Filter is activated at the same time.

and

> Clicking on the Expand/Collapse Glyph expands/collapses the second area of the Filter, where the Input Fields are placed. When the Filter Bar is expanded, the row with set filter settings at the bottom of the Filter Bar will be hidden. Clicking on the "Toggle" Button imitates a click on the "Apply" Bulky Button, activates/deactivates the Filter and reloads the content of the item collection immediately.

However, there seems to be several issue here (all token from the example, note that it is very well possible, that I did not get something right):

- If active, the filter can not be collapsed (which makes it also impossible the check the „Additional Row with set filter settings“, shown at the bottom of the Filter bar, whan collapsed and activated).
- When toggled to inactive, the set items are not removed, just deactivated (which is good). However if collapsed, and then shown again, all items disappear (including default ones). This is imo not good. Why is that?

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [x] Make use of the „use“ statement for namesapces. This would make some parts more readable (see inline comment on that)
- [x] Why does collapse and expand trigger a full page reload, maybe revise to simply hide and show by JS. Am I missing something?

Please implement the following changes:
- [ ] Close Popovers if other field is clicked, currently they stay open.
- [ ] Fix the collapse/expand issue(s) (see questions above).
- [ ] Resolve remaining conflics


@klees promised to deliver his review until the end of the week. Do not hesitate to contact us, if there are questions left. Please ping us if you think this is ready for merging.

kindly,
@Amstutz

klees commented on src/UI/Implementation/Component/Input/Container/Filter/Filter.php:
```diff
@@ -0,0 +1,343 @@
+	 * @param string|Signal $apply_action
+	 * @param string|Signal $reset_action
+	 * @param array $inputs
+	 * @param array $is_input_rendered
```
```suggestion
	 * @param bool[] $is_input_rendered
```

klees commented on src/UI/Implementation/Component/Input/Container/Filter/Filter.php:
```diff
@@ -0,0 +1,343 @@
+	 * @param string|Signal $collapse_action
+	 * @param string|Signal $apply_action
+	 * @param string|Signal $reset_action
+	 * @param array $inputs
```
```suggestion
	 * @param C\Input\Field\Input[] $inputs
```

klees commented on src/UI/Implementation/Component/Input/Container/Filter/Filter.php:
```diff
@@ -0,0 +1,343 @@
+
+	// Implementation of NameSource
+
+	public function getNewName() {
```
```suggestion
	/**
	 * @inheritdoc
	 */
	public function getNewName() {
```

klees commented on src/UI/Implementation/Component/Input/Container/Filter/PostDataFromServerRequest.php:
```diff
@@ -0,0 +1,50 @@
+	 */
+	public function get($name) {
+		if (!isset($this->query_params[$name])) {
+			throw new \LogicException("'$name' is not contained in posted data.");
```
```suggestion
			throw new \LogicException("'$name' is not contained in query parameters.");
```

klees commented on src/UI/templates/js/Popover/popover.js:
```diff
@@ -73,12 +73,26 @@ il.UI = il.UI || {};
          */
         var show = function($triggerer, options) {
+
+            //$('#' + $triggerer.attr('id')).trigger("il.ui.popover.before-show");
```
```suggestion
```

klees reviewed (changes requested):
Hi @tfamula!

Thanks a lot for working so hard on the filter! I definitely see that we are making progress here and I am also sorry that my review brings up such a long list of stuff. I still do not consider it an option to just be silent about stuff I dislike here, instead I'm sure we will find a way to solve the issues one way or another. I'm also sure that you'll make your way through (this review and the filter stuff). You seem to be on a good track to become a very knowledgeable UI-person here =)

I concentrated my review on the UI-framework part of the PR and only skimmed over the `Services\UI`-part since we already agreed that we might reconsider this part at some point when the UI-framework knows more about control structure.

It would have helped a lot if you had rebased the PR to trunk (or merged the trunk), because it seems to contain code that already is included via #1305. I did not review the JS-code so far because the odd indentation makes this extremely hard atm. Please ping me when you have fixed the indentation and I will have a look there.

Please answer the following questions:
- [ ] Why do we need both, `isInputRendered` (and according parameter) and `isDisabled` on the inputs? How are the two related? Why would I want to add a disabled input to the filter? If I remember correctly (this might of course be wrong and I also cannot find documentation about that) we identified the disabled-functionality for input fields as a way to get rid of the oddly disconnected boolean-list that tells which fields needs to be rendered and which don't. Using the disabled-property of the fields themself would directly attach that boolean to the element it belongs to and nicely unify the two concepts "disabled inputs in forms" and "fields to filter". Since this is not implemented, however, I feel that there will be some notewothy explanation here (or I simply misremember).
- [x] [This line here](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-aa6d1db0578554908359daddf8f13f16R223) contains a question from yourself (I assume).
- [ ] What is the purpose of `ProxyFilterField`? It caught my attention because it is only created once in a single location and is directly consumed afterwards. This to me suggests that there is some functionality you want to access there but couldn't access in another way, which in turn could hint at some possible improvement.
- [x] What is the purpose of the change in `tagInput.js`?

Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
- [x] Remove [this](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-a4ae795be99f4117745c8d359752deccR41) comment, looks like a copy and paste remainder.
- [x] Please inject dependencies [here](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-aa6d1db0578554908359daddf8f13f16R121) instead of creating them directly. I think this should be easy since #1054 is merged.
- [x] Rename `PostDataFromServerRequest` to `QueryParamsFromServerRequest`, since that's what essentially used here, correct? Consider renaming `PostData` as well, I would suggest to `InputData`. We certainly do not only use POST here after your change. If this is too much effort please add an according entry to the roadmap.
- [x] `Input\Container\Filter\Renderer::renderExpand` seems to contain duplication. Please refactor it to remove the duplicated code. Also the name seems to be misleading, "collapse" is renderered as well.
- [x] Rename `GroupHelper::withUpdateSignal` to `withOnUpdate` and let `Group` implement `OnUpdateable`. For one, it is hard to understand why `withUpdateSignal` takes a `Component` as argument (and not a Signal). It also is not documented which components can be processed. Furthermore I cannot see a reason why this should no be `withOnUpdate`, instead the implementation suggests that this is exactly that.
- [ ] Do not deactivate the input group contained in the filter [in the Renderer](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-08841937cae3b8ab3f50c5b841fc93f9R239) but [in the component itself](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-aa6d1db0578554908359daddf8f13f16R286) instead. Consider removing the internal `is_activated`-property in the Filter-component completely in favour of falling back to the "disabled"-property of the contained group. You could still keep the according `with*` and `get*` on the filter, though. This will remove one point where things could be in an odd state or forgotten.

Please implement the following changes:
- [ ] Replace `getUpdateOnLoadCode` and replace it by using the new OnUpdate-signal. If I remember correctly (and @Amstutz comment from 11-27 suggests so as well) we wanted to get rid of the `FilterInput`-interface, which only contained the `getUpdateOnLoadCode` at the time we discussed this. Instead we decided to introduce the OnUpdate-signal. This signal now exists, but `getUpdateOnLoad`-code (and documentation) exists as well, while the OnUpdate-signal is not used for the filter. Also the name for `getUpdateOnLoadCode` is in fact lying. It is not some general on load code for updates but instead very specific code for the filter, which is exactly what we wanted to improve with the OnUpdate-signal.
- [ ] Do not add [empty methods](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-48267845b56fb4a2b62352167f358f9cR478) that are defined on an interface that is supposed to be implemented. The interfaces reminds people to implement the missing method. Also, this will throw once called, since we hint on `\Closure`.
- [x] Please remove the `else` and unindent the code contained in [this block here](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-aa6d1db0578554908359daddf8f13f16R116). If the code runs into the `if` it will throw anyway and the parts afterwards will never be reached. Also, only some parts of the following code are contained in the block without an obvious reason.
- [x] Remove uncommented code. Git will keep it for us if we really need it.
- [x] The indendation in `src/UI/templates/js/Input/Container/filter.js` is off in many places.

I understand that you want to move on with this and I certainly also want to have the filters in the trunk. However, I still see problems where at least one problem (`getUpdateOnLoadCode`) has already been adressed in our last meeting and is not solved yet. On the other hand I see a lot of cleanup that needs to be done in this PR, where some of it already made the review harder than it must have been. If you are exhausted by this project or PR and cannot implement some (or even all) of my suggestions and change requests we could think about transforming them to entries of the roadmap so we could at least keep track of them.

If you have any questions about my review feel free to contact me.

kindly,
@klees

tfamula committed:
Filter improvements, part 1

tfamula committed:
Filter improvements, part 2

tfamula commented:
Hi @Amstutz and @klees! Thank you for your elaborate feedback. In recent days, I was tackling to understand your comments and to implement the open points. Some of them are finished and others are partly/not finished and need discussion. To have a better overview, I quoted your reviews before adding my comments:

> Hi @tfamula
> Thanks a lot for your great stamina on this topic. This is complex however, we are getting there.
> My review is mainly based on some basic checks I performed on your example, making sure, the behaviour of the implementation, matches the given the description. @rklees might leave some more comments about the more technical aspects of this.
> 
> Please answer the following questions:
> 
> * [ ]  What is the relation between collapse and activate? From your description I gather:
> 
> > An additional row with set filter settings will be shown at the bottom of the Filter Bar, when the Filter Bar is collapsed and the Filter is activated at the same time.
> 
> and
> 
> > Clicking on the Expand/Collapse Glyph expands/collapses the second area of the Filter, where the Input Fields are placed. When the Filter Bar is expanded, the row with set filter settings at the bottom of the Filter Bar will be hidden. Clicking on the "Toggle" Button imitates a click on the "Apply" Bulky Button, activates/deactivates the Filter and reloads the content of the item collection immediately.
> 
> However, there seems to be several issue here (all token from the example, note that it is very well possible, that I did not get something right):
> 
> * If active, the filter can not be collapsed (which makes it also impossible the check the „Additional Row with set filter settings“, shown at the bottom of the Filter bar, whan collapsed and activated).

We see your point here. The rules indeed describe the Filter differently as one sees it, at least in the desktop view. In the desktop view, a Filter **is not** collapsible when active. In the mobile, a Filter **is** collapsible when active. The reason for this is [a decision from the JF](https://docu.ilias.de/goto_docu_wiki_wpage_3687_1357.html#ilPageTocA152) on 18 JUN 2018. It recommends “to merge the two steps 'Activate' and 'Show' for the desktop. If a user activates the filter, the filter elements should be shown, too.” Our understanding from this was to show the filter always expanded in the desktop view when it is activated. In the mobile view, this should not be like that due to space reasons. We could either modify the rules or abandon this behaviour. We are okay with both options and have to find a decision here with you.

> * When toggled to inactive, the set items are not removed, just deactivated (which is good). However if collapsed, and then shown again, all items disappear (including default ones). This is imo not good. Why is that?

Thank you for detecting this bug, it has been fixed.

> 
> Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
> 
> * [x]  Make use of the „use“ statement for namesapces. This would make some parts more readable (see inline comment on that)
> * [ ]  Why does collapse and expand trigger a full page reload, maybe revise to simply hide and show by JS. Am I missing something?

On the very beginning, I implemented it like that when the expand/collapse state were not defined by the consumer and when they were not actions yet. Currently, they are actions. An action means a new request must be triggered.
An alternative could be to change it to an Ajax Request. This would cause inconsistency. In total, we have six actions (toggle on, toggle off, apply, reset, expand, collapse). The other four actions cannot be transformed to Ajax Requests because they update content on the page and therefore need a page reload. After changing two actions to Ajax, we would have multiple solutions at the same time.

> 
> Please implement the following changes:
> 
> * [ ]  Close Popovers if other field is clicked, currently they stay open.

We think that this is not a bug of the Filter but from the Popover itself, because this effect also appears in every other place using the Popover. If this behaviour is unwanted and you expect a similar behaviour like in Dropdowns, this should be fixed in the Popover element in our opinion.

> * [ ]  Fix the collapse/expand issue(s) (see questions above).

See above

> * [ ]  Resolve remaining conflics

This will be done as a last step before merging to trunk. The fails and errors in the tests will be fixed in the end, too.

> 
> @klees promised to deliver his review until the end of the week. Do not hesitate to contact us, if there are questions left. Please ping us if you think this is ready for merging.
> 
> kindly,
> @Amstutz





> Hi @tfamula!
> 
> Thanks a lot for working so hard on the filter! I definitely see that we are making progress here and I am also sorry that my review brings up such a long list of stuff. I still do not consider it an option to just be silent about stuff I dislike here, instead I'm sure we will find a way to solve the issues one way or another. I'm also sure that you'll make your way through (this review and the filter stuff). You seem to be on a good track to become a very knowledgeable UI-person here =)
> 
> I concentrated my review on the UI-framework part of the PR and only skimmed over the `Services\UI`-part since we already agreed that we might reconsider this part at some point when the UI-framework knows more about control structure.
> 
> It would have helped a lot if you had rebased the PR to trunk (or merged the trunk), because it seems to contain code that already is included via #1305. I did not review the JS-code so far because the odd indentation makes this extremely hard atm. Please ping me when you have fixed the indentation and I will have a look there.

It contains code from #1305 because of cherry-picks. They were necessary to make Filters work with Disabled Inputs. In our experience rebasing is messing up the PRs (at least this happened in the past). And we would like to avoid the need to switch to a new PR and lose the comments.

> 
> Please answer the following questions:
> 
> * [ ]  Why do we need both, `isInputRendered` (and according parameter) and `isDisabled` on the inputs? How are the two related? Why would I want to add a disabled input to the filter? If I remember correctly (this might of course be wrong and I also cannot find documentation about that) we identified the disabled-functionality for input fields as a way to get rid of the oddly disconnected boolean-list that tells which fields needs to be rendered and which don't. Using the disabled-property of the fields themself would directly attach that boolean to the element it belongs to and nicely unify the two concepts "disabled inputs in forms" and "fields to filter". Since this is not implemented, however, I feel that there will be some notewothy explanation here (or I simply misremember).

`isInputRendered` and `isDisabled` are different things. I also do not remember that we wanted to get rid of the first method by replacing it with the second method.
With `isInputRendered`, the consumer defines which Inputs should be shown as Input Fields in the Filter to directly operate with them and which Inputs should be stored in the “Add” Bulky Button to add them when needed. Moreover, the parameters from `isInputRendered` are used in the Filter Service to handle the correct rendering of the Filter Inputs after interacting with the Filter.
`isDisabled` was introduced for all Inputs (not only Inputs which can be used in the Filter currently). It can be used in Forms and Filters and prevents users to interact with Inputs for some reason. In case of Filters, Disabled Inputs are used when the Filter is deactivated (Toggle Button is off). But `isDisabled` does not decide if an Input Field will be shown in the Filter or lies hidden in the “Add” Bulky Button. The user sees the same Input Fields like in the acitvated Filter state. 

> * [x]  [This line here](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-aa6d1db0578554908359daddf8f13f16R223) contains a question from yourself (I assume).
> * [ ]  What is the purpose of `ProxyFilterField`? It caught my attention because it is only created once in a single location and is directly consumed afterwards. This to me suggests that there is some functionality you want to access there but couldn't access in another way, which in turn could hint at some possible improvement.

This has already been discussed/explained [here](https://github.com/ILIAS-eLearning/ILIAS/pull/1193#pullrequestreview-158064039). The class `ProxyFilterField` is used to make the Input Field in the Filter (which is not the “real” input) clickable with `withOnClick`.

> * [x]  What is the purpose of the change in `tagInput.js`?

See above. I cherry-picked the changes from the Disabled Inputs PR into this PR to make appropriate changes for the Filter.

> 
> Please consider the following suggestions. You do not need to follow those, but but please indicate shortly why you prefer to do otherwise:
> 
> * [x]  Remove [this](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-a4ae795be99f4117745c8d359752deccR41) comment, looks like a copy and paste remainder.
> * [x]  Please inject dependencies [here](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-aa6d1db0578554908359daddf8f13f16R121) instead of creating them directly. I think this should be easy since #1054 is merged.
> * [x]  Rename `PostDataFromServerRequest` to `QueryParamsFromServerRequest`, since that's what essentially used here, correct? Consider renaming `PostData` as well, I would suggest to `InputData`. We certainly do not only use POST here after your change. If this is too much effort please add an according entry to the roadmap.
> * [x]  `Input\Container\Filter\Renderer::renderExpand` seems to contain duplication. Please refactor it to remove the duplicated code. Also the name seems to be misleading, "collapse" is renderered as well.

`renderExpand` was renamed to `renderExpandOrCollapse`. Found little duplication removed (most code here is similar but not a duplication IMO)

> * [x]  Rename `GroupHelper::withUpdateSignal` to `withOnUpdate` and let `Group` implement `OnUpdateable`. For one, it is hard to understand why `withUpdateSignal` takes a `Component` as argument (and not a Signal). It also is not documented which components can be processed. Furthermore I cannot see a reason why this should no be `withOnUpdate`, instead the implementation suggests that this is exactly that.

Thank you for this hint. This was indeed more complicated than needed. But Group does not need to implement `OnUpdateable` because it extends Input and Input does it already.

> * [ ]  Do not deactivate the input group contained in the filter [in the Renderer](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-08841937cae3b8ab3f50c5b841fc93f9R239) but [in the component itself](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-aa6d1db0578554908359daddf8f13f16R286) instead. Consider removing the internal `is_activated`-property in the Filter-component completely in favour of falling back to the "disabled"-property of the contained group. You could still keep the according `with*` and `get*` on the filter, though. This will remove one point where things could be in an odd state or forgotten.

I could not find a way to do this. I can make the input group deactivated in the Component instead of the Renderer, but this will only work when the `withDeacitvated` method is used explicitly for the Filter. Without using the method, the activation status can only be checked in the `Renderer`, where the decision is made how Inputs will be rendered in the `FilterContextRenderer`.
Moreover, the `is_activated` property of the Filter and the disabled property of the Inputs should be handled separately. If a consumer wants to use a disabled Input in an activated Filter for some good reason, this should be possible.

> 
> Please implement the following changes:
> 
> * [ ]  Replace `getUpdateOnLoadCode` and replace it by using the new OnUpdate-signal. If I remember correctly (and @Amstutz comment from 11-27 suggests so as well) we wanted to get rid of the `FilterInput`-interface, which only contained the `getUpdateOnLoadCode` at the time we discussed this. Instead we decided to introduce the OnUpdate-signal. This signal now exists, but `getUpdateOnLoad`-code (and documentation) exists as well, while the OnUpdate-signal is not used for the filter. Also the name for `getUpdateOnLoadCode` is in fact lying. It is not some general on load code for updates but instead very specific code for the filter, which is exactly what we wanted to improve with the OnUpdate-signal.

The decision from 11-27 was to introduce the `OnUpdate`-signal for all Inputs. This has been done here. The signal can be used now for every Input, regardless whether used in Forms or Filters.
It is correct that we talked about getting rid of the `FilterInput`-interface. Unfortunately, this is not possible at the moment, because we have to ensure that only Inputs are used in the Filter which are compatible for it by synchronizing the Input in the Popover with the Input in the Filter. Getting rid of the `FilterInput`-Interface could be achieved by forcing the developer to implement the `getUpdateOnLoadCode`-method (see comment in next point below).
The `OnUpdate`-signals are only part of the interface between the component and the signal consumer (here `FilterInput` and `Filter`).
But additionally we need to ensure that `FilterInput` components implement Javascript that binds changes in the HTML input elements to triggering the OnUpdate signals in the first place. We could move this easily to the renderer like the Button component does (see [Button Renderer](https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/src/UI/Implementation/Component/Button/Renderer.php#L73). But this would hide the obligation for `FilterInput` developers instead of making it explicit in an interface.

> * [ ]  Do not add [empty methods](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-48267845b56fb4a2b62352167f358f9cR478) that are defined on an interface that is supposed to be implemented. The interfaces reminds people to implement the missing method. Also, this will throw once called, since we hint on `\Closure`.

We could declare the method itself as abstract in the Input class and try to implement it for each existing Input. Since this would make some effort, we would like to wait until the OnUpdate-signal stuff mentioned before is resolved.

> * [x]  Please remove the `else` and unindent the code contained in [this block here](https://github.com/ILIAS-eLearning/ILIAS/pull/1253/files#diff-aa6d1db0578554908359daddf8f13f16R116). If the code runs into the `if` it will throw anyway and the parts afterwards will never be reached. Also, only some parts of the following code are contained in the block without an obvious reason.
> * [x]  Remove uncommented code. Git will keep it for us if we really need it.

Removed to the best of my knowledge and belief

> * [x]  The indendation in `src/UI/templates/js/Input/Container/filter.js` is off in many places.
> 
> I understand that you want to move on with this and I certainly also want to have the filters in the trunk. However, I still see problems where at least one problem (`getUpdateOnLoadCode`) has already been adressed in our last meeting and is not solved yet. On the other hand I see a lot of cleanup that needs to be done in this PR, where some of it already made the review harder than it must have been. If you are exhausted by this project or PR and cannot implement some (or even all) of my suggestions and change requests we could think about transforming them to entries of the roadmap so we could at least keep track of them.
> 
> If you have any questions about my review feel free to contact me.
> 
> kindly,
> @klees


As you can see there are several open points to discuss. To make the process faster, we would suggest to resolve them in a VC as soon as you are ready. Especially, we should find a decision about:

- how to deal with the collapsible/not collapsible behaviour in mobile/desktop view

- collapse and expand actions (full page reload due to new request)

- the `OnUpdate`-signal and `getUpdateOnLoadCode`-method

tfamula committed:
Small fixes

Amstutz commented:
We (@klees, @tfamula, @alex40724 and @amstutz) discussed the above listed issues today march 13th 2019. We agreed on the following:

- The comment of the JF in [Responsive Table Filters](https://docu.ilias.de/goto_docu_wiki_wpage_3687_1357.html) on the 18th of June "We suggest to merge the two steps 'Activate' and 'Show' for the desktop. If a user activates the filter, the filter elements should be shown, too. This improves usability.“ can be understood in different ways.
- We agreed that we propose as simplified alternative to treat desktop and small screen size in regard to collapse expand and active/inactive in the same way. That is: allow expand and collapse in both states of the filters, activated and deactivated, on desktop and mobile. We will announce this at the next JF and propose the changes to the JF.
- To handle the problem with the full round-trip on expand we will use ajax-calls using the base-url with exit in the service.
- The getUpdateOnLoadCode problem will be tackled by making them abstract and requiring ALL Inputs to implement them and add the JS in each rendering of each input. However we will only tackle it, if there are resources left for the filter project after tackling the above issues. If not, we will add an according Input in the roadmap.

Thanks to @klees, @tfamula and @alex40724 for the productive meeting.

tfamula committed:
Changes for expand/collapse Filter with Ajax

tfamula committed:
Bulky Button instead of Glyph in Filter Bar

tfamula committed:
Slight border to separate Filter Bar from Inputs

matthiaskunkel commented:
Jour Fixe, 25 MAR 2019 : We accept the change compared to our decision from June 2018 concerning the behaviour of expand and activation of the filter on desktop. Behaviour should be similar on mobile and desktop. Expand and actvation should be separated. Using AJAX calls instead of page reloads is fine for us. We also accept the new function getUpdateOnLoadCode for every input.

tfamula committed:
Design fixes for very long Filter Inputs

Amstutz commented:
To avoid any confusion, Ill close this down. The most recent one is: #1735.
